File: add-challenges.js
// add-challenges.js
// Script to add new challenges to MongoDB

const mongoose = require('mongoose');
require('dotenv').config();
const challengesData = require('./challenges-data');

// Connect to MongoDB
const connectDB = async () => {
  try {
    // Get MongoDB URI from environment variable
    const mongoURI = process.env.MONGO_URI;
    if (!mongoURI) {
      console.error('MONGO_URI environment variable is not set');
      return false;
    }

    // Connect to MongoDB
    await mongoose.connect(mongoURI, {
      useNewUrlParser: true,
      useUnifiedTopology: true
    });
    console.log('MongoDB connected successfully');
    return true;
  } catch (error) {
    console.error('MongoDB connection error:', error);
    return false;
  }
};

// Define Challenge schema directly in this script
// This ensures we don't need to rely on the model file
const challengeSchema = new mongoose.Schema({
  title: { type: String, required: true },
  description: { type: String, required: true },
  points: { type: Number, default: 10 },
  thumbnail: { type: String, required: true },
  type: { type: String, default: 'practice_video' },
  createdAt: { type: Date, default: Date.now },
  active: { type: Boolean, default: true }
});

// Add challenges to database
const addChallenges = async () => {
  try {
    // Connect to database
    const connected = await connectDB();
    if (!connected) {
      console.error('Failed to connect to MongoDB. Exiting...');
      process.exit(1);
    }

    // Get the database name from the connection string
    const dbName = 'fpt_learning'; // Use the same database name as in app.py

    // Get the challenges collection
    const db = mongoose.connection.db;
    const challengesCollection = db.collection('challenges');

    console.log(`Adding challenges to database ${dbName}...`);

    // Track results
    let inserted = 0;
    let updated = 0;
    let errors = 0;

    // Process each challenge
    for (const challenge of challengesData) {
      try {
        // Add createdAt if not present
        if (!challenge.createdAt) {
          challenge.createdAt = new Date();
        }

        // Ensure active flag is set
        if (challenge.active === undefined) {
          challenge.active = true;
        }

        // Check if challenge with same title already exists
        const existingChallenge = await challengesCollection.findOne({ title: challenge.title });

        if (existingChallenge) {
          // Update existing challenge
          const result = await challengesCollection.updateOne(
            { title: challenge.title },
            { $set: challenge }
          );

          if (result.modifiedCount > 0) {
            console.log(`Updated challenge: ${challenge.title}`);
            updated++;
          } else {
            console.log(`No changes needed for: ${challenge.title}`);
          }
        } else {
          // Insert new challenge
          const result = await challengesCollection.insertOne(challenge);
          console.log(`Added new challenge: ${challenge.title}`);
          inserted++;
        }
      } catch (err) {
        console.error(`Error processing challenge "${challenge.title}":`, err);
        errors++;
      }
    }

    // Summary
    console.log('\nSummary:');
    console.log(`- ${inserted} challenges inserted`);
    console.log(`- ${updated} challenges updated`);
    console.log(`- ${errors} errors encountered`);

    // Disconnect from database
    await mongoose.disconnect();
    console.log('Disconnected from MongoDB');

  } catch (error) {
    console.error('Error adding challenges:', error);
  }
};

// Run the function
addChallenges();

File: add_challenges.py
#!/usr/bin/env python3
# add_challenges.py
# Script to add new challenges to MongoDB

import os
import json
from datetime import datetime, timezone
from dotenv import load_dotenv
from pymongo import MongoClient

# Load environment variables
load_dotenv()

# Get MongoDB URI from environment variable
MONGO_URI = os.getenv("MONGO_URI")
if not MONGO_URI:
    print("Error: MONGO_URI environment variable not set.")
    exit(1)

# Define challenges data
challenges_data = [
    {
        "title": "Thử Thách Sáo Trúc: Bài Hát Dân Ca",
        "description": "Quay video thể hiện kỹ năng thổi sáo trúc với một bài hát dân ca Việt Nam tự chọn. Thời lượng tối thiểu 1 phút.",
        "points": 30,
        "thumbnail": "/assets/images/challenges/challenge_folk_song.jpg",
        "type": "practice_video",
        "active": True
    },
    {
        "title": "Thử Thách Đàn Tranh: Kỹ Thuật Rung Dây",
        "description": "Thực hiện và quay video kỹ thuật rung dây đàn tranh với 3 mức độ khác nhau: nhẹ, vừa và mạnh.",
        "points": 25,
        "thumbnail": "/assets/images/challenges/challenge_dan_tranh.jpg",
        "type": "practice_video",
        "active": True
    },
    {
        "title": "Thử Thách Vovinam: Đòn Chân Tấn Công",
        "description": "Thực hiện chính xác và quay video kỹ thuật đòn chân tấn công trong Vovinam. Chú ý đến tư thế, góc đá và lực.",
        "points": 35,
        "thumbnail": "/assets/images/challenges/challenge_vovinam_kick.jpg",
        "type": "practice_video",
        "active": True
    },
    {
        "title": "Thử Thách Lý Thuyết Âm Nhạc",
        "description": "Hoàn thành bài kiểm tra 15 câu hỏi về lý thuyết âm nhạc cơ bản, bao gồm các nốt nhạc, nhịp điệu và hòa âm.",
        "points": 20,
        "thumbnail": "/assets/images/challenges/challenge_music_theory.jpg",
        "type": "quiz",
        "active": True
    },
    {
        "title": "Thử Thách Sáng Tạo: Biến Tấu Giai Điệu",
        "description": "Tạo một biến tấu sáng tạo từ một giai điệu truyền thống và quay video trình diễn. Hãy giải thích cách bạn thay đổi giai điệu gốc.",
        "points": 40,
        "thumbnail": "/assets/images/challenges/challenge_creative.jpg",
        "type": "practice_video",
        "active": True
    },
    {
        "title": "Thử Thách Võ Thuật: Bài Quyền Cơ Bản",
        "description": "Thực hiện và quay video một bài quyền cơ bản trong võ thuật bạn đang học. Chú ý đến tư thế, nhịp điệu và sự chính xác của các động tác.",
        "points": 30,
        "thumbnail": "/assets/images/challenges/challenge_martial_form.jpg",
        "type": "practice_video",
        "active": True
    },
    {
        "title": "Thử Thách Flashcards: Thuật Ngữ Chuyên Ngành",
        "description": "Hoàn thành bài kiểm tra 20 flashcards về thuật ngữ chuyên ngành trong lĩnh vực bạn đang học (nhạc cụ hoặc võ thuật).",
        "points": 25,
        "thumbnail": "/assets/images/challenges/challenge_terminology.jpg",
        "type": "flashcard_test",
        "active": True
    }
]

def add_challenges():
    try:
        # Connect to MongoDB
        print(f"Connecting to MongoDB...")
        client = MongoClient(MONGO_URI, serverSelectionTimeoutMS=5000)
        client.admin.command('ping')  # Test connection
        db = client['fpt_learning']  # Use the specific DB name
        print(f"Connected to MongoDB. Using database: {db.name}")

        # Track results
        inserted = 0
        updated = 0
        errors = 0

        # Process each challenge
        print("Adding challenges to database...")
        for challenge in challenges_data:
            try:
                # Add createdAt if not present
                if 'createdAt' not in challenge:
                    challenge['createdAt'] = datetime.now(timezone.utc)

                # Check if challenge with same title already exists
                existing_challenge = db.challenges.find_one({'title': challenge['title']})

                if existing_challenge:
                    # Update existing challenge
                    result = db.challenges.update_one(
                        {'title': challenge['title']},
                        {'$set': challenge}
                    )

                    if result.modified_count > 0:
                        print(f"Updated challenge: {challenge['title']}")
                        updated += 1
                    else:
                        print(f"No changes needed for: {challenge['title']}")
                else:
                    # Insert new challenge
                    result = db.challenges.insert_one(challenge)
                    print(f"Added new challenge: {challenge['title']}")
                    inserted += 1
            except Exception as err:
                print(f"Error processing challenge '{challenge['title']}': {err}")
                errors += 1

        # Summary
        print("\nSummary:")
        print(f"- {inserted} challenges inserted")
        print(f"- {updated} challenges updated")
        print(f"- {errors} errors encountered")

    except Exception as e:
        print(f"Error connecting to MongoDB: {e}")
    finally:
        # Close the connection
        if 'client' in locals():
            client.close()
            print("Disconnected from MongoDB")

if __name__ == "__main__":
    add_challenges()
